### DISCUSSION TITLE
Apollo cannot start stream with dual GPU setup, any idea?

### DISCUSSION THREAD

**Discussion by kohanyirobert** [ClassicOldSong/Apollo#41] [General]
Hey! 👋 I've recently got into game streaming, found Sunshine, but since it wasn't able to handle my dual GPU laptop (ASUS ROG Zephyrus G14 2022) I've started looking into alternatives and found Apollo. For my use case both could work but since the README states

> If you want to use your dGPU, just set the Adapter Name to your dGPU and enable Headless mode in Audio/Video tab, save and restart your computer. No dummy plug is needed any more, the image will be rendered and encoded directly from your dGPU.

I thought I'll give it a shot. To no avail.

My adapters are an AMD Radeon(TM) Graphics (iGPU) and AMD Radeon RX 6800S (dGPU) and whatever I do I cannot get Apollo to work consistently Currently, I restarted my machine and "flipped" the MUX switch in the laptop so that everything is routed to the dGPU by default, but still I get the following

```
[2024-09-17 19:15:24.237]: Info: Virtual Display created at \\.\DISPLAY12
[2024-09-17 19:15:24.249]: Info: // Testing for available encoders, this may generate errors. You can safely ignore those errors. //
[2024-09-17 19:15:24.249]: Info: Trying encoder [nvenc]
[2024-09-17 19:15:24.765]: Error: Failed to locate an output device
[2024-09-17 19:15:25.267]: Error: Failed to locate an output device
[2024-09-17 19:15:25.984]: Error: Failed to locate an output device
[2024-09-17 19:15:26.493]: Error: Failed to locate an output device
[2024-09-17 19:15:26.707]: Info: Encoder [nvenc] failed
[2024-09-17 19:15:26.707]: Info: Trying encoder [quicksync]
[2024-09-17 19:15:27.211]: Error: Failed to locate an output device
[2024-09-17 19:15:27.725]: Error: Failed to locate an output device
[2024-09-17 19:15:28.442]: Error: Failed to locate an output device
[2024-09-17 19:15:28.948]: Error: Failed to locate an output device
[2024-09-17 19:15:29.152]: Info: Encoder [quicksync] failed
[2024-09-17 19:15:29.152]: Info: Trying encoder [amdvce]
[2024-09-17 19:15:29.654]: Error: Failed to locate an output device
[2024-09-17 19:15:30.162]: Error: Failed to locate an output device
[2024-09-17 19:15:30.877]: Error: Failed to locate an output device
[2024-09-17 19:15:31.387]: Error: Failed to locate an output device
[2024-09-17 19:15:31.594]: Info: Encoder [amdvce] failed
[2024-09-17 19:15:31.595]: Info: Trying encoder [software]
[2024-09-17 19:15:32.102]: Error: Failed to locate an output device
[2024-09-17 19:15:32.611]: Error: Failed to locate an output device
[2024-09-17 19:15:33.322]: Error: Failed to locate an output device
[2024-09-17 19:15:33.826]: Error: Failed to locate an output device
[2024-09-17 19:15:34.031]: Info: Encoder [software] failed
[2024-09-17 19:15:34.031]: Fatal: Unable to find display or encoder during startup.
[2024-09-17 19:15:34.031]: Fatal: Please ensure your manually chosen GPU and monitor are connected and powered on.
[2024-09-17 19:15:34.161]: Info: Virtual Display removed successfully
[2024-09-17 19:16:04.013]: Info: Virtual Display created at \\.\DISPLAY12
```

however if I disable my iGPU driver in the Device Manager it works. Still not optimal.

Wanted to start this discussion to see how others deal with this, whether there's another solution, etc.

Thanks chipping in.

**Comment by ClassicOldSong** [comment]
Can you post a screen shot of your Audio/Video tab and display adapters section in your device manager? Does it work without headless mode and left the adapter name config blank?

**Reply by kohanyirobert** [reply]
Everything is left at its default except the adapter name which I copied from the output of dxgi-info.exe and enabling headless mode.

I'll try without headless soon and report back, but I believe I tried already and the result was the same.

![image](https://github.com/user-attachments/assets/07e3b369-665a-4b95-9c24-68a5a7958538)

**Reply by kohanyirobert** [reply]
First of all, I'm using Apollo v0.1.4-alpha.2. Using Artemis on Android as the client (but had similar issues with Moonlight as well).

OK, so here's what I tried. 
1. I reenabled my iGPU in device manager (still in MUX switch _on_ mode - I'm using G-Helper to manage this, so I'm Ultimate mode if that's any help for anyone reading this)
2. Disabled headless mode
3. Restarted Apollo
4. Connected with both the original "Desktop" and "Virtual Desktop" app configurations, both worked.

Now here comes the interesting part.
5. Because headless mode (based on the description) only makes every app connect to a virtual display I _re-enabled_ it
6. Still both Desktop and Virtual Desktop worked

Have no idea why, so it's pretty inconsistent, like before connecting to a non-virtual desktop it just cannot work. Hope it makes sense. I'll test it some more but I think have to be more rigorous, like differentiating between "first starts", having my Ultimate mode enabled or not,  etc.

Hope any of this helps.

**Reply by ClassicOldSong** [reply]
Now check dxgi-info.exe and see which graphics card the virtual display actually connects to

**Reply by ClassicOldSong** [reply]
Also if you could post the screen shots of dxgi-info.exe under different configurations(mux on/off, iGPU disabled/enabled), that would also help.

**Reply by kohanyirobert** [reply]
Will do, it takes some time to prepare for this.

On a side note: I've noticed that I have two driver entry for my dGPU in device manager having slightly different device instance path IDs or whatever (available in device manager driver details view). Not sure if important. Maybe it's related to having my MUX switched on or off, I'll take this into account and report back. Let's get to the bottom of this :)

**Reply by kohanyirobert** [reply]
I'm back with the results, I've tested various configs. It turns out the extra device in the device manager cannot be disabled or enabled and it's not used when switching between GPU profiles.

https://docs.google.com/spreadsheets/d/192oMBrazZ2KbWdsRWer_C9BzZrNBdO0LR65UoKrkthE

(Hope this helps, let me know if you can't decipher something.)

I've collected general info about my setup on the General tab/sheet of the spreadsheet, and Scenarios shows which config worked: to sum up nothing worked expect switching my GPU profile to Ultimate (display is routed to dGPU via MUX switch) _and_ disabling my iGPU device.

I didn't date disable my iGPU device when my display is routed to this device, but probably that would make thing work also (having a single device helps).

**Reply by ClassicOldSong** [reply]
I haven't seen any configurations like this before. I'm not sure what's going on but it might be related to those two entries. Still, info from dxgi-info.exe is important.

IDK if this helps, but you can try enabling headless mode, leave adapter name and display name empty, save the config, then run `regedit`, add an entry named `gpuName` in `HKEY_LOCAL_MACHINE\SOFTWARE\SudoMaker\SudoVDA` with type `String`, set its value to `AMD Radeon RX 6800S`, then reboot your computer or disable and then re-enable `SudoMaker Virtual Display Adapter` in device manager. You need to quit Apollo before doing these.

Then start Apollo and see if it helps.

**Reply by ClassicOldSong** [reply]
Also upgrade to the latest version, it might help since probing has improved a bit.

**Reply by ClassicOldSong** [reply]
Seems related:

https://community.amd.com/t5/pc-drivers-software/need-help-to-enable-amd-h264-h65-encoding-on-680m-rx-6500m/td-p/594771

Unfortunately this configuration might not be able to encode with hardware with both iGPU and dGPU enabled at the same time...

**Reply by kohanyirobert** [reply]
dxgi-info.exe output, forgot about that. This is the output when I'm in "Standard" mode, display routed to iGPU and dGPU enabled.

```
====== ADAPTER =====
Device Name      : AMD Radeon(TM) Graphics
Device Vendor ID : 0x00001002
Device Device ID : 0x00001681
Device Video Mem : 483 MiB
Device Sys Mem   : 0 MiB
Share Sys Mem    : 15990 MiB

    ====== OUTPUT ======
    Output Name       : \\.\DISPLAY1
    AttachedToDesktop : yes
    Resolution        : 1920x1080

====== ADAPTER =====
Device Name      : AMD Radeon RX 6800S
Device Vendor ID : 0x00001002
Device Device ID : 0x000073EF
Device Video Mem : 8146 MiB
Device Sys Mem   : 0 MiB
Share Sys Mem    : 15990 MiB

    ====== OUTPUT ======
====== ADAPTER =====
Device Name      : AMD Radeon(TM) Graphics
Device Vendor ID : 0x00001002
Device Device ID : 0x00001681
Device Video Mem : 483 MiB
Device Sys Mem   : 0 MiB
Share Sys Mem    : 15990 MiB

    ====== OUTPUT ======
====== ADAPTER =====
Device Name      : Microsoft Basic Render Driver
Device Vendor ID : 0x00001414
Device Device ID : 0x0000008C
Device Video Mem : 0 MiB
Device Sys Mem   : 0 MiB
Share Sys Mem    : 15990 MiB

    ====== OUTPUT ======
```

**Reply by kohanyirobert** [reply]
> Seems related:
> 
> https://community.amd.com/t5/pc-drivers-software/need-help-to-enable-amd-h264-h65-encoding-on-680m-rx-6500m/td-p/594771
> 
> Unfortunately this configuration might not be able to encode with hardware with both iGPU and dGPU enabled at the same time...

Well, I'm neither an expert in GPUs at all, nor in dual GPU setups, but I'm not convinced by this thread (only one guy's posts :S). Not saying you aren't right just this evidence isn't substantial.

Will try the regedit trick, then update to the latest version too. Thanks for the tips.

**Reply by kohanyirobert** [reply]
Hmm, using regedit as suggested did the trick at first look.

I'll do a few restarts and test it some more tho'!

**Reply by kohanyirobert** [reply]
dxgi-info after I successfully connect using headless mode, doing the regedit trick and having my two GPUs enabled (no fiddling with them).

```
====== ADAPTER =====
Device Name      : AMD Radeon RX 6800S
Device Vendor ID : 0x00001002
Device Device ID : 0x000073EF
Device Video Mem : 8146 MiB
Device Sys Mem   : 0 MiB
Share Sys Mem    : 15990 MiB

    ====== OUTPUT ======
    Output Name       : \\.\DISPLAY32
    AttachedToDesktop : yes
    Resolution        : 2312x1080

====== ADAPTER =====
Device Name      : AMD Radeon(TM) Graphics
Device Vendor ID : 0x00001002
Device Device ID : 0x00001681
Device Video Mem : 483 MiB
Device Sys Mem   : 0 MiB
Share Sys Mem    : 15990 MiB

    ====== OUTPUT ======
====== ADAPTER =====
Device Name      : AMD Radeon RX 6800S
Device Vendor ID : 0x00001002
Device Device ID : 0x000073EF
Device Video Mem : 8146 MiB
Device Sys Mem   : 0 MiB
Share Sys Mem    : 15990 MiB

    ====== OUTPUT ======
====== ADAPTER =====
Device Name      : Microsoft Basic Render Driver
Device Vendor ID : 0x00001414
Device Device ID : 0x0000008C
Device Video Mem : 0 MiB
Device Sys Mem   : 0 MiB
Share Sys Mem    : 15990 MiB

    ====== OUTPUT ======
```

**Reply by kohanyirobert** [reply]
It seems it works, even after I restart I'm consistently able to connect to my laptop.

Btw. I didn't update Apollo, only did the regedit trick.

Thanks for help mate!

**Reply by ClassicOldSong** [reply]
It's still recommended to upgrade, the new version contains more bug fixes and the new permission system is really nice.

**Reply by kohanyirobert** [reply]
Will do eventually. The permission system is a non-issue for me.
The killer feature of Apollo over Sunshine is virtual display management. Apollo enables/disables this automatically, if this works, I'm set.

**Reply by meloblerd** [reply]
@ClassicOldSong This discussion has helped me get this working. On my Asus G14 GA402XY. Regarding the Ultimate mode though. This routes the dGPU to the laptop screen which gives the best performance. Ideally we can get this to work in Standard mode where the laptop can have both dGPU and iGPU enabled. Which is what is implied by dual GPU.

**Comment by kohanyirobert** [comment]
Well, still no good. Now I experience the following:
1. Both iGPU + dGPU are enabled
2. Launching Apollo in headless mode, adding my adapter's name in regedit
3. Starting Virtual Desktop _connects_ but the stream is black.

This is the dxgi-info.exe output _while_ connected to the stream. And I'm attaching the log also.

It seems the correct adapter is used, but in the logs there's no mention of Radeon RX 6800S, only the iGPU's adapter name is there. Tried setting the adapter name after this in Apollo, but then I get the 503 error on the client, the stream doesn't even start. Not sure what made it work yesterday ...

[sunshine.log](https://github.com/user-attachments/files/17083224/sunshine.log)

```
====== ADAPTER =====
Device Name      : AMD Radeon RX 6800S
Device Vendor ID : 0x00001002
Device Device ID : 0x000073EF
Device Video Mem : 8146 MiB
Device Sys Mem   : 0 MiB
Share Sys Mem    : 15990 MiB

    ====== OUTPUT ======
    Output Name       : \\.\DISPLAY12
    AttachedToDesktop : yes
    Resolution        : 2312x1080

====== ADAPTER =====
Device Name      : AMD Radeon(TM) Graphics
Device Vendor ID : 0x00001002
Device Device ID : 0x00001681
Device Video Mem : 483 MiB
Device Sys Mem   : 0 MiB
Share Sys Mem    : 15990 MiB

    ====== OUTPUT ======
====== ADAPTER =====
Device Name      : AMD Radeon RX 6800S
Device Vendor ID : 0x00001002
Device Device ID : 0x000073EF
Device Video Mem : 8146 MiB
Device Sys Mem   : 0 MiB
Share Sys Mem    : 15990 MiB

    ====== OUTPUT ======
====== ADAPTER =====
Device Name      : Microsoft Basic Render Driver
Device Vendor ID : 0x00001414
Device Device ID : 0x0000008C
Device Video Mem : 0 MiB
Device Sys Mem   : 0 MiB
Share Sys Mem    : 15990 MiB

    ====== OUTPUT ======
```

**Reply by ClassicOldSong** [reply]
It seems that this can only be solved if I can physically get my hands on one of these devices...

Too weird for having one card showing two entries...

**Reply by ClassicOldSong** [reply]
One last thing you can try: manually set sunshine.exe and ddprobe.exe to "High Performance" in Windows settings

**Reply by ClassicOldSong** [reply]
Looked at your log and found something weird. Would you try this build and see if it works?

https://github.com/ClassicOldSong/Apollo/releases/tag/v0.2.2-alpha.2

**Reply by kohanyirobert** [reply]
> One last thing you can try: manually set sunshine.exe and ddprobe.exe to "High Performance" in Windows settings
> Looked at your log and found something weird. Would you try this build and see if it works?

Will try both when it breaks again, currently the previous config I reported working (then not working) works again 😄 ... my hunch is that somehow somewhere Windows or some subsystem keeps track of what GPU was last used for what screens (external or not) or similar and only in some constellation of states/settings it works.

I'll keep an eye on this.
